inet.man
.SH INET ADDRESS FAMILY
This section documents the methods available in the
inet address family.
.SS The loopback Method
This method may be used to define the IPv4 loopback interface. 
.PP
.B Options
.RS
.TP
(No options)
.RE
.SS The static Method
This method may be used to define Ethernet interfaces with statically allocated IPv4 addresses. 
.PP
.B Options
.RS
.TP
.BI address " address            "
Address (dotted quad/netmask) \fBrequired\fP
.TP
.BI netmask " mask               "
Netmask (dotted quad or CIDR)
.TP
.BI broadcast " broadcast_address"
Broadcast address (dotted quad, + or -). Default value: "+"
.TP
.BI metric " metric              "
Routing metric for default gateway (integer)
.TP
.BI gateway " address            "
Default gateway (dotted quad)
.TP
.BI pointopoint " address        "
Address of other end point (dotted quad). Note the spelling of "point-to".
.TP
.BI hwaddress " address          "
Link local address or "random".
.TP
.BI mtu " size                   "
MTU size
.TP
.BI scope
Address validity scope. Possible values: global, link, host
.RE
.SS The manual Method
This method may be used to define interfaces for which no configuration is done by default. Such interfaces can be configured manually by means of \fBup\fP and \fBdown\fP commands or /etc/network/if-*.d scripts. 
.PP
.B Options
.RS
.TP
.BI hwaddress " address          "
Link local address or "random".
.TP
.BI mtu " size                   "
MTU size
.RE
.SS The dhcp Method
This method may be used to obtain an address via DHCP with any of the tools: dhclient, pump, udhcpc, dhcpcd. (They have been listed in their order of precedence.) If you have a complicated DHCP setup you should note that some of these clients use their own configuration files and do not obtain their configuration information via \fBifup\fP. 
.PP
.B Options
.RS
.TP
.BI hostname " hostname      "
Hostname to be requested (pump, dhcpcd, udhcpc)
.TP
.BI metric " metric          "
Metric for added routes (dhclient)
.TP
.BI leasehours " leasehours  "
Preferred lease time in hours (pump)
.TP
.BI leasetime " leasetime    "
Preferred lease time in seconds (dhcpcd)
.TP
.BI vendor " vendor          "
Vendor class identifier (dhcpcd)
.TP
.BI client " client          "
Client identifier (dhcpcd)
.TP
.BI hwaddress " address      "
Hardware address.
.RE
.SS The bootp Method
This method may be used to obtain an address via bootp. 
.PP
.B Options
.RS
.TP
.BI bootfile " file "
Tell the server to use \fIfile\fP as the bootfile.
.TP
.BI server " address"
Use the IP address \fIaddress\fP to communicate with the server.
.TP
.BI hwaddr " addr   "
Use \fIaddr\fP as the hardware address instead of whatever it really is.
.RE
.SS The tunnel Method
This method is used to create GRE or IPIP tunnels. You need to have the \fBip\fP binary from the \fBiproute\fP package. For GRE tunnels, you will need to load the ip_gre module and the ipip module for IPIP tunnels. 
.PP
.B Options
.RS
.TP
.BI address " address      "
Local address (dotted quad) \fBrequired\fP
.TP
.BI mode " type            "
Tunnel type (either GRE or IPIP) \fBrequired\fP
.TP
.BI endpoint " address     "
Address of other tunnel endpoint \fBrequired\fP
.TP
.BI dstaddr " address      "
Remote address (remote address inside tunnel)
.TP
.BI local " address        "
Address of the local endpoint
.TP
.BI metric " metric        "
Routing metric for default gateway (integer)
.TP
.BI gateway " address      "
Default gateway
.TP
.BI ttl " time             "
TTL setting
.TP
.BI mtu " size             "
MTU size
.RE
.SS The ppp Method
This method uses pon/poff to configure a PPP interface. See those commands for details. 
.PP
.B Options
.RS
.TP
.BI provider " name "
Use \fIname\fP as the provider (from /etc/ppp/peers).
.TP
.BI unit " number   "
Use \fInumber\fP as the ppp unit number.
.TP
.BI options " string"
Pass \fIstring\fP as additional options to pon.
.RE
.SS The wvdial Method
This method uses wvdial to configure a PPP interface. See that command for more details. 
.PP
.B Options
.RS
.TP
.BI provider " name "
Use \fIname\fP as the provider (from /etc/wvdial.conf).
.RE
.SS The ipv4ll Method
This method uses avahi-autoipd to configure an interface with an IPv4 Link-Layer address (169.254.0.0/16 family). This method is also known as APIPA or IPAC, and often colloquially referred to as "Zeroconf address". 
.PP
.B Options
.RS
.TP
(No options)
.RE


debian/changelog
ifupdown (0.8.19+2) unstable; urgency=medium

  * ifupdown-hotplug changes
  * networking to shut down interfaces at shutdown 

 -- anticapitalista <antix@operamail.com>  Tue, 03 Oct 2017 13:27:02 +0300

ifupdown (0.8.19+1) unstable; urgency=medium

  * antiX build 

 -- anticapitalista <antix@operamail.com>  Wed, 20 Sep 2017 00:24:45 +0100

ifupdown (0.8.19) unstable; urgency=medium

  * Fix exit code of ifquery --state. Closes: #852976

 -- Guus Sliepen <guus@debian.org>  Mon, 30 Jan 2017 14:32:18 +0100

ifupdown (0.8.18) unstable; urgency=medium

  [ Svante Signell ]
  * Makefile: Add removal of tests/*/state.* directories in the clean target.
  * Update testcases for GNU/Hurd.

  [ Guus Sliepen ]
  * Use dynamically allocated strings for filenames. This avoids using
    PATH_MAX, which caused a FTBFS on Hurd. Based on Svante's patch.
    Closes: #845140
  * Fix some memory leaks.
  * Improve error messages.
  * Always immediately quit when running out of memory.

 -- Guus Sliepen <guus@debian.org>  Wed, 11 Jan 2017 20:43:06 +0100

ifupdown (0.8.17) unstable; urgency=medium

  [ Scott Moser ]
  * Don't bring down the loopback interface when networking.service stops.
    Closes: #841106, #841107

  [ Guus Sliepen ]
  * Export $CLASS to scripts.
  * Add a bug-script.
  * Depend on freebsd-net-tools instead of net-tools.
  * Also configure VLANs on bridge interfaces (see #818849).
  * Improve manpage: simplify example, add more section headers.
    Closes: #840579
  * Provide IPv6 examples in /u/s/d/ifupdown/examples/network-interfaces.
    Closes: #703568
  * Use -4 option for IPv4 dhclient. Closes: #844592

  [ Martin Pitt ]
  * ifup@.service: start After the corresponding
    sys-subsystem-net-devices device.

 -- Guus Sliepen <guus@debian.org>  Tue, 10 Jan 2017 17:25:18 +0100

ifupdown (0.8.16) unstable; urgency=medium

  * Fix FTBFS due to testcases using the host's state information.

 -- Guus Sliepen <guus@debian.org>  Thu, 22 Sep 2016 13:20:18 +0200

ifupdown (0.8.15) unstable; urgency=medium

  * Fix ifdown complaining about vlan_id1. Closes: #838312
  * Add testcases for ifdown. 

 -- Guus Sliepen <guus@debian.org>  Thu, 22 Sep 2016 11:53:52 +0200

ifupdown (0.8.14) unstable; urgency=medium

  [ Benjamin Drung ]
  * Add InfiniBand parititon key support. Closes: #830212

  [ Richard Laager ]
  * Fix a crash when multiple interfaces are specified for no-scripts.
    Closes: #837732

  [ Bjørn Mork ]
  * Ignore link state when bringing up hotplug interfaces at boot.
    Closes: #814785, #834820

  [ Guus Sliepen ]
  * Add a test case for the no-scripts keyword.
  * Flush automatically assigned addresses on ifup when accept_ra=0.

 -- Guus Sliepen <guus@debian.org>  Sat, 17 Sep 2016 11:40:39 +0200

ifupdown (0.8.13) unstable; urgency=medium

  * Fix test suite on kFreeBSD. Closes: #826062
  * Check the return value of chdir() and ftruncate().
  * Clarify that scripts can be run with IFACE set to "--all".
    Closes: #826000

 -- Guus Sliepen <guus@debian.org>  Sat, 04 Jun 2016 14:08:54 +0200

ifupdown (0.8.12) unstable; urgency=medium

  * Add "nodad" to ip -6 addr add if dad-attempts is set to zero.
    Closes: #824686
  * Bump Standards-Version.

 -- Guus Sliepen <guus@debian.org>  Thu, 19 May 2016 15:36:02 +0200

ifupdown (0.8.11) unstable; urgency=medium

  [ Wido den Hollander ]
  * Wait properly for Link-Local Address to go through DAD.

  [ Imre Deak ]
  * Fix read_all_state when no state file exists. Closes: 819287

  [ Arthur Gautier ]
  * Implement link naming (ip link set alias)

  [ Martin Pitt <martin.pitt@ubuntu.com> ]
  * Add autopkgtest for "allow-hotplug" interfaces. Closes: #814312

  [ Guus Sliepen ]
  * Bump Standards-Version.

 -- Guus Sliepen <guus@debian.org>  Thu, 21 Apr 2016 18:22:20 +0200

ifupdown (0.8.10) unstable; urgency=medium

  * Allow "hwaddress random" to generate a random MAC address.
  * Allow ifdown/ifquery when physical interface is not same as logical
    interface. 

 -- Guus Sliepen <guus@debian.org>  Fri, 22 Jan 2016 16:35:23 +0100

ifupdown (0.8.9) unstable; urgency=medium

  * It's network-online.target, not .service.
  * Do not run "ip link set down" on manual interfaces during "ifdown -a".
    Closes: #809166
  * Add the keywords "no-auto-down" and "no-scripts". Closes: #615130

 -- Guus Sliepen <guus@debian.org>  Thu, 21 Jan 2016 22:02:11 +0100

ifupdown (0.8.8) unstable; urgency=medium

  [ Martin Pitt ]
  * Fix ifquery crash if interface state file does not exist yet.
    (Closes: #810779, LP: #1532722)
  * ifup@.service: Avoid stopping on shutdown via stopping system-ifup.slice
    (changed behaviour in systemd 228). (Closes: #761909, LP: #1492546)

  [ Guus Sliepen ]
  * Remove quotes around TimeoutStartSec parameter. Closes: #810656

 -- Guus Sliepen <guus@debian.org>  Tue, 12 Jan 2016 23:50:29 +0100

ifupdown (0.8.7) unstable; urgency=medium

  * Impose a 5 minute timeout for starting network services.
    Closes: #318577, #810656
  * Give an error when trying to up a dhcp interface when no DHCP client is
    installed. Closes: #471926, #419139
  * Add a section in the manual about extensions provided by other packages.
    Closes: #482405, #597274
  * Use absolute paths to binaries called by ifup/ifdown. Closes: #365114
  * Return an error when trying to ifup/ifdown/ifquery an unknown interface
    (unless it is ignored due to --allow or --exclude).
  * Ignore SIGPIPE. Closes: #311054
  * Update the bridge example. Closes: #488315
  * Use the onlink option when adding a gateway route. Closes: #378506
  * Also allow the metric option for tunnels and IPv6 interfaces.

 -- Guus Sliepen <guus@debian.org>  Mon, 11 Jan 2016 22:58:44 +0100

ifupdown (0.8.6) unstable; urgency=medium

  * Remove the Breaks: udev (<< 228-3~) to work around a bug in udev's prerm
    script. Closes: #809658, #809743
  * Run ip link up on manual interfaces, but ignore any errors.
  * Use sysctl to set the IPv6-specific MTU of an interface. Closes: #809714


inet.c
#include <stddef.h>
#include "header.h"


#include "archcommon.h"
#include "archlinux.h"


static option_default _loopback_default[] = {
	{ NULL, NULL }
};
static conversion _loopback_conv[] = {
	{ NULL, NULL, NULL, 0, NULL }
};
static int _loopback_up(interface_defn *ifd, execfn *exec) {
if ( !iface_is_lo() ) {
  if (!execute("/bin/ip link set dev %iface% up", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _loopback_down(interface_defn *ifd, execfn *exec) {
if ( !iface_is_lo() ) {
  if (!execute("/bin/ip link set dev %iface% down", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static option_default _static_default[] = {
	{ "broadcast", "+" },
	{ NULL, NULL }
};
static conversion _static_conv[] = {
	{ "hwaddress", NULL, cleanup_hwaddress, 0, NULL },
	{ "address", "netmask?", compute_v4_mask, 0, NULL },
	{ "address", NULL, compute_v4_addr, 0, NULL },
	{ "broadcast", NULL, compute_v4_broadcast, 0, NULL },
	{ NULL, NULL, NULL, 0, NULL }
};
static int _static_up(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/bin/ip addr add %address%[[/%netmask%]] [[broadcast %broadcast%]] \
	[[peer %pointopoint%]] [[scope %scope%]] dev %iface% label %iface%", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip link set dev %iface% [[mtu %mtu%]] [[address %hwaddress%]] up", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("[[ /bin/ip route add default via %gateway% [[metric %metric%]] dev %iface% onlink ]]", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _static_down(interface_defn *ifd, execfn *exec) {
{
  if (!execute("[[ /bin/ip route del default via %gateway% [[metric %metric%]] dev %iface% 2>&1 1>/dev/null || true ]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip addr del %address%[[/%netmask%]] [[broadcast %broadcast%]] \
	[[peer %pointopoint%]] [[scope %scope%]] dev %iface% label %iface%", ifd, exec) && !ignore_failures) return 0;
}
if ( iface_is_link() ) {
  if (!execute("/bin/ip link set dev %iface% down", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static option_default _manual_default[] = {
	{ NULL, NULL }
};
static conversion _manual_conv[] = {
	{ "hwaddress", NULL, cleanup_hwaddress, 0, NULL },
	{ NULL, NULL, NULL, 0, NULL }
};
static int _manual_up(interface_defn *ifd, execfn *exec) {
{
  if (!execute("[[/bin/ip link set dev %iface% mtu %mtu%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("[[/bin/ip link set dev %iface% address %hwaddress%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip link set dev %iface% up 2>/dev/null || true", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _manual_down(interface_defn *ifd, execfn *exec) {
if ( iface_is_link() && !do_all ) {
  if (!execute("/bin/ip link set dev %iface% down 2>/dev/null || true", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static option_default _dhcp_default[] = {
	{ NULL, NULL }
};
static conversion _dhcp_conv[] = {
	{ "hwaddress", NULL, cleanup_hwaddress, 0, NULL },
	{ NULL, NULL, NULL, 0, NULL }
};
static int _dhcp_up(interface_defn *ifd, execfn *exec) {
{
  if (!execute("[[/bin/ip link set dev %iface% address %hwaddress%]]", ifd, exec) && !ignore_failures) return 0;
}
if ( execable("/sbin/dhclient") ) {
  if (!execute("/sbin/dhclient -4 -v -pf /run/dhclient.%iface%.pid -lf /var/lib/dhcp/dhclient.%iface%.leases -I -df /var/lib/dhcp/dhclient6.%iface%.leases %iface% \
	[[-e IF_METRIC=%metric%]]", ifd, exec) && !ignore_failures) return 0;
}
else if ( execable("/sbin/pump") ) {
  if (!execute("/sbin/pump -i %iface% [[-h %hostname%]] [[-l %leasehours%]]", ifd, exec) && !ignore_failures) return 0;
}
else if ( execable("/sbin/udhcpc") ) {
  if (!execute("/sbin/udhcpc -n -p /run/udhcpc.%iface%.pid -i %iface% [[-x hostname:%hostname%]]", ifd, exec) && !ignore_failures) return 0;
}
else if ( execable("/sbin/dhcpcd") ) {
  if (!execute("/sbin/dhcpcd [[-h %hostname%]] [[-i %vendor%]] [[-I %client%]] \
           [[-l %leasetime%]] [[-m %metric%]] %iface%", ifd, exec) && !ignore_failures) return 0;
}
else if ( 1 ) {
  if (!execute("echo 'No DHCP client software found!' >/dev/stderr; false", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _dhcp_down(interface_defn *ifd, execfn *exec) {
if ( execable("/sbin/dhclient") ) {
  if (!execute("/sbin/dhclient -4 -v -r -pf /run/dhclient.%iface%.pid -lf /var/lib/dhcp/dhclient.%iface%.leases -I -df /var/lib/dhcp/dhclient6.%iface%.leases %iface%", ifd, exec) && !ignore_failures) return 0;
}
else if ( execable("/sbin/pump") ) {
  if (!execute("/sbin/pump -i %iface% -r", ifd, exec) && !ignore_failures) return 0;
}
else if ( execable("/sbin/udhcpc") ) {
  if (!execute("if test -f /run/udhcpc.%iface%.pid; then kill -USR2 $(/bin/cat /run/udhcpc.%iface%.pid); kill -TERM $(/bin/cat /run/udhcpc.%iface%.pid); fi", ifd, exec) && !ignore_failures) return 0;
}
else if ( execable("/sbin/dhcpcd") ) {
  if (!execute("/sbin/dhcpcd -k %iface%", ifd, exec) && !ignore_failures) return 0;
}
else if ( 1 ) {
  if (!execute("echo 'No DHCP client software found!' >/dev/stderr; false", ifd, exec) && !ignore_failures) return 0;
}
if ( iface_is_link() ) {
  if (!execute("/bin/ip link set dev %iface% down", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static option_default _bootp_default[] = {
	{ NULL, NULL }
};
static conversion _bootp_conv[] = {
	{ NULL, NULL, NULL, 0, NULL }
};
static int _bootp_up(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/sbin/bootpc [[--bootfile %bootfile%]] --dev %iface% [[--server %server%]] \
           [[--hwaddr %hwaddr%]] --returniffail --serverbcast", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _bootp_down(interface_defn *ifd, execfn *exec) {
if ( iface_is_link() ) {
  if (!execute("/bin/ip link set dev %iface% down", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static option_default _tunnel_default[] = {
	{ NULL, NULL }
};
static conversion _tunnel_conv[] = {
	{ NULL, NULL, NULL, 0, NULL }
};
static int _tunnel_up(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/bin/ip tunnel add %iface% mode %mode% remote %endpoint% [[local %local%]] \
       [[ttl %ttl%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip link set %iface% up [[mtu %mtu%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("/bin/ip addr add %address%/%netmask% dev %iface% [[peer %dstaddr%]]", ifd, exec) && !ignore_failures) return 0;
}
{
  if (!execute("[[ /bin/ip route add default via %gateway% [[metric %metric%]] dev %iface% onlink ]]", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _tunnel_down(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/bin/ip tunnel del %iface%", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static option_default _ppp_default[] = {
	{ NULL, NULL }
};
static conversion _ppp_conv[] = {
	{ NULL, NULL, NULL, 0, NULL }
};
static int _ppp_up(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/usr/bin/pon [[%provider%]] [[unit %unit%]] [[%options%]]", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _ppp_down(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/usr/bin/poff [[%provider%]]", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static option_default _wvdial_default[] = {
	{ NULL, NULL }
};
static conversion _wvdial_conv[] = {
	{ NULL, NULL, NULL, 0, NULL }
};
static int _wvdial_up(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/sbin/start-stop-daemon --start -x /usr/bin/wvdial \
                      -p /run/wvdial.%iface%.pid -b -m -- [[ %provider% ]]", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _wvdial_down(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/sbin/start-stop-daemon --stop -x /usr/bin/wvdial \
                      -p /run/wvdial.%iface%.pid -s 2", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static option_default _ipv4ll_default[] = {
	{ NULL, NULL }
};
static conversion _ipv4ll_conv[] = {
	{ NULL, NULL, NULL, 0, NULL }
};
static int _ipv4ll_up(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/usr/sbin/avahi-autoipd -D %iface%", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static int _ipv4ll_down(interface_defn *ifd, execfn *exec) {
{
  if (!execute("/usr/sbin/avahi-autoipd --kill %iface%", ifd, exec) && !ignore_failures) return 0;
}
return 1;
}
static method methods[] = {
        {
                "bootp",
                _bootp_up, _bootp_down,
                _bootp_conv, _bootp_default
        },
        {
                "dhcp",
                _dhcp_up, _dhcp_down,
                _dhcp_conv, _dhcp_default
        },
        {
                "ipv4ll",
                _ipv4ll_up, _ipv4ll_down,
                _ipv4ll_conv, _ipv4ll_default
        },
        {
                "loopback",
                _loopback_up, _loopback_down,
                _loopback_conv, _loopback_default
        },
        {
                "manual",
                _manual_up, _manual_down,
                _manual_conv, _manual_default
        },
        {
                "ppp",
                _ppp_up, _ppp_down,
                _ppp_conv, _ppp_default
        },
        {
                "static",
                _static_up, _static_down,
                _static_conv, _static_default
        },
        {
                "tunnel",
                _tunnel_up, _tunnel_down,
                _tunnel_conv, _tunnel_default
        },
        {
                "wvdial",
                _wvdial_up, _wvdial_down,
                _wvdial_conv, _wvdial_default
        },
};

address_family addr_inet = {
        "inet",
        sizeof(methods)/sizeof(struct method),
        methods
};
